File: Day278_2025/Q278_Leetcode.py
from typing import List

class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        n, m = len(heights), len(heights[0])

        pac = [[False] * m for _ in range(n)]
        atl = [[False] * m for _ in range(n)]

        directions = [(1,0), (-1,0), (0,1), (0,-1)]

        def dfs(r, c, vis, prev_height):
            if (r < 0 or r >= n or c < 0 or c >= m or
                vis[r][c] or heights[r][c] < prev_height):
                return
            vis[r][c] = True
            for dr, dc in directions:
                dfs(r + dr, c + dc, vis, heights[r][c])

        # Pacific (top row + left col)
        for i in range(n):
            dfs(i, 0, pac, -1)
            dfs(i, m - 1, atl, -1)
        for j in range(m):
            dfs(0, j, pac, -1)
            dfs(n - 1, j, atl, -1)

        ans = []
        for i in range(n):
            for j in range(m):
                if pac[i][j] and atl[i][j]:
                    ans.append([i, j])
        return ans

File: Day278_2025/Q278_Leetcode.java
import java.util.*;

class Q278_Leetcode {
    private int n, m;
    private int[][] heights;
    private boolean[][] pac, atl;

    private void dfs(int i, int j, boolean[][] vis, int prevHeight) {
        if (i < 0 || i >= n || j < 0 || j >= m || vis[i][j] || heights[i][j] < prevHeight)
            return;
        vis[i][j] = true;
        dfs(i - 1, j, vis, heights[i][j]);
        dfs(i + 1, j, vis, heights[i][j]);
        dfs(i, j - 1, vis, heights[i][j]);
        dfs(i, j + 1, vis, heights[i][j]);
    }

    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        this.n = heights.length;
        this.m = heights[0].length;
        this.heights = heights;
        pac = new boolean[n][m];
        atl = new boolean[n][m];

        // Pacific (left + top borders)
        for (int i = 0; i < n; i++) {
            dfs(i, 0, pac, -1);
            dfs(i, m - 1, atl, -1);
        }
        for (int j = 0; j < m; j++) {
            dfs(0, j, pac, -1);
            dfs(n - 1, j, atl, -1);
        }

        List<List<Integer>> ans = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (pac[i][j] && atl[i][j]) {
                    ans.add(Arrays.asList(i, j));
                }
            }
        }
        return ans;
    }
}
